pymc_experimental/distributions/continuous.py

```diff
@@ -0,0 +1,166 @@
+        alpha = at.as_tensor_variable(floatX(alpha))
+        p = at.as_tensor_variable(floatX(p))
+        lambd = at.as_tensor_variable(floatX(lambd))
+
+        return super().dist([alpha, p, lambd], **kwargs)
+
+    def moment(rv, size, alpha, p, lambd):
+        alpha, p, lambd = at.broadcast_arrays(alpha, p, lambd)
+        moment = lambd * at.gamma((alpha + 1) / p) / at.gamma(alpha / p)
+        if not rv_size_is_none(size):
+            moment = at.full(size, moment)
+        return moment
+
+    def logp(
+        value,
+        alpha: Union[float, np.ndarray, TensorVariable],
+        p: Union[float, np.ndarray, TensorVariable],
+        lambd: Union[float, np.ndarray, TensorVariable],
+    ) -> RandomVariable:
+        """
+        Calculate log-probability of Generalized Gamma distribution at specified value.
+        Parameters
+        ----------
+        value : tensor_like of float
+            Value(s) for which log-probability is calculated. If the log probabilities for multiple
+            values are desired the values must be provided in a numpy array or Aesara tensor.
+        alpha : tensor_like of float
+            Shape parameter (alpha > 0).
+        p : tensor_like of float
+            Shape parameter (p > 0).
+        lambd : tensor_like of float
+            Scale parameter (lambd > 0).
+        Returns
+        -------
+        TensorVariable
+        """
+        logp_expression = (
+            at.log(p)
+            - at.log(lambd)
+            + (alpha - 1) * at.log(value / lambd)
+            - (value / lambd) ** p
+            - at.math.gammaln(alpha / p)
+        )
+
+        bounded_logp_expression = at.switch(
+            at.gt(value, 0),
+            logp_expression,
+            -np.inf,
+        )
+
+        return check_parameters(
+            bounded_logp_expression,
+            alpha > 0,
+            p > 0,
+            lambd > 0,
+            msg="alpha > 0, p > 0, lambd > 0",
+        )
+
+    def logcdf(
+        value,
+        alpha: Union[float, np.ndarray, TensorVariable],
+        p: Union[float, np.ndarray, TensorVariable],
+        lambd: Union[float, np.ndarray, TensorVariable],
+    ) -> RandomVariable:
+        """
+        Compute the log of the cumulative distribution function for GeneralizedGamma
+        distribution at the specified value.
+        Parameters
+        ----------
+        value : tensor_like of float
+            Value(s) for which log CDF is calculated. If the log CDF for multiple
+            values are desired the values must be provided in a numpy array or Aesara tensor.
+        alpha : tensor_like of float
+            Shape parameter (alpha > 0).
+        p : tensor_like of float
+            Shape parameter (p > 0).
+        lambd : tensor_like of float
+            Scale parameter (lambd > 0).
+        Returns
+        -------
+        TensorVariable
+        """
+        logcdf_expression = at.log(at.gammainc(alpha / p, (value / lambd) ** p))
```

Comment:
I'm not very familiar with the math, but I am concerned the logcdf function _may_ have some numerical instability. I attempted to fit a model to simulated data with the `pm.Censored` API and it was a bit slow, and had a few divergences

 (I tested this with the [latest aeppl fix](https://github.com/aesara-devs/aeppl/pull/156) for the `pm.Censored` API).

Reply:
I don't see some obvious reparametrization. Unless there is some loggammaincc function we could use?

If you are concerned I would suggest you do a grid evaluation of the logcdf and evaluate against an arbitrary precision software like Mathematica.

That will confirm or dispell your concerns of stability, and then we can try to look for an alternative. It could also be the instability is not in the function itself but its gradients.

Do you have a minimum example where you found the sampling issues that you can share?

Reply:
Is Mathematica a paid software? I had trouble finding anything to test against that would guarantee numerical precision

here's a minimum (kind of) example where there are sampling issues (very slow sampling. The distribution is quick when not using the `pm.Censored` api)

```
import pandas as pd
import scipy.stats as stats
import pymc as pm
import arviz as az
import numpy as np
import matplotlib.pyplot as plt
import pymc_experimental as pmx


SEED = 99


# ########################
# code
# ########################

def sim_data(true_paramsA, true_paramsB, SEED):
	
	np.random.seed(SEED)

	# 2 groups that generate data
	a = stats.gengamma(true_paramsA[0]/true_paramsA[1], true_paramsA[1], scale=true_paramsA[2])
	b = stats.gengamma(true_paramsB[0]/true_paramsB[1], true_paramsB[1], scale=true_paramsB[2])
	N = 100 # obs per group

	# create a dataset of N obs from each group
	y_true = np.r_[
	        a.rvs(N),
	        b.rvs(N)
	    ]

	# randomly censor the dataset for survival analysis
	cens_time = np.random.uniform(0, y_true.max(), size=N*2)

	data = (
		pd.DataFrame({
		"group":[0]*N + [1]*N,
		"time": y_true})
		# adjust the dataset to censor observations
		## indicates if an event hasnt occurred yet (cens=1)
		.assign(cens = lambda d: np.where(d.time <= cens_time, 0, 1) )
		## indicates the latest time observed for each record
		.assign(time = lambda d: np.where(d.cens==1, cens_time, d.time) )
	)

	return data

def fit_model(data):

	cens_ = np.where(data.cens==1, data.time, np.inf)
	COORDS = {"group": ["A","B"]}
	group_ = data.group.values
	y_ = data.time.values

	with pm.Model(coords=COORDS) as model:

		# weakly informative priors for each parameter
		log_alpha = pm.Normal("log_alpha", 2, 0.75, dims="group")
		log_p = pm.Normal("log_p", 0.45, 0.3, dims="group")
		log_lambd = pm.Normal("log_lambd", 4.1, 0.4, dims="group")

		# helper vars
		alpha = pm.Deterministic("alpha", pm.math.exp(log_alpha), dims="group")
		p = pm.Deterministic("p", pm.math.exp(log_p), dims="group")
		lambd = pm.Deterministic("lambd", pm.math.exp(log_lambd), dims="group")

		# latent variable needed for censored fitting
		y_latent = pmx.GeneralizedGamma.dist( alpha[group_], p[group_], lambd[group_] )

		# likelihood of the observed data
		obs = pm.Censored("obs", y_latent, lower=None, upper=cens_, observed=y_)

		# fit
		idata = pm.sample(init="adapt_diag", random_seed=SEED)
		idata.extend(pm.sample_prior_predictive())
		idata.extend(pm.sample_posterior_predictive(idata))

	return model, idata


def main():
	true_paramsA = [2, 2, 60]
	true_paramsB = [20, 2, 60]
	data = sim_data(true_paramsA, true_paramsB, SEED)
	model, idata = fit_model(data)


if __name__ == '__main__':
	
	main()
```

Reply:
Are you using the main version of Aeppl? We haven't done a new release yet, but there was a recent patch that makes the gradients of censored logps more stable: https://github.com/aesara-devs/aeppl/pull/156

Reply:
> Is Mathematica a paid software?

I might have jumbled the names. In the past I've used https://www.wolframcloud.com/ with a free account.

Reply:
Thank you I'll take a look at that!

I'm currently using the main version of aeppl installed manually (with `pip install -e .` inside of a locally cloned aeppl repo), so it should have the censored logp update. The sampling is >10 minutes to fit a dataset of 200 observations, but it takes less than a minute if I dont use the `Censored` api

Some of the warnings I'm getting are:

```
/Users/kylejcaron/.pyenv/versions/3.9.7/lib/python3.9/site-packages/aesara/scalar/math.py:760: RuntimeWarning: invalid value encountered in double_scalars
  scipy.special.gammaincc(k, x) * (log_x - digamma_k)
```

and 

```
WARNING (aesara.tensor.basic_opt): Optimization Warning: The Op gammainc_der does not provide a C implementation. As well as being potentially slow, this also disables loop fusion.
```

Reply:
The other thing I just noticed is that theres correlation between the `lambda` parameter and the `p` parameter. Are there any suggested ways to try and reduce that correlation? Maybe I simply just have to try other parameterizations?

I'm not seeing anything too crazy in the trace plots however

![download-1](https://user-images.githubusercontent.com/44980552/183137014-91cfc924-e5d5-435d-a0a6-0ca123559b64.png)
![download](https://user-images.githubusercontent.com/44980552/183137016-4d996813-f88a-4889-8ffa-bdca41bbf603.png)

Reply:
The gradient using a Python loop could definitely explain slowness :P

But perhaps the censored model is just difficult to sample? Does it improve if you reduce the amount of censoring?

Is a censored normal much faster?

Reply:
One thing to note is that everything is fine and samples quickly without censored data and without the `censored` api involved (i.e. just normal fitting on non-censored data)

Also, sampling might've actually gotten slower with less censoring (when using the `censored` api)? I didnt sample to completion, I just looked at the initial guess after 1 minute of sampling
 * Original, 46% (96/200 obs censored) - 40 minutes of sampling estimated
 * 25% censoring (50/200 censored) - 50 minutes
 * 2.5% (5/200 censored) - 1 hour 

 * Censored normal data sampled quickly for me (<5 seconds)
 * Censored Gamma was really slow (estimated to be > 1 hour. Also got slower with less censoring). Uncensored fitting on same data was fine.

The fact that the fitting gets slower with less censoring seems really off to me, and also surprised to see the Gamma distribution suffers from similar issues

Reply:
update: just looked at the [log-cdf of the gamma distribution](https://github.com/pymc-devs/pymc/blob/e64942be0bc643b2ffb709213546f3d0243bb5f7/pymc/distributions/continuous.py#L2320) and its implemented very similar to how I implemented the generalized gamma, so makes sense its slow - I guess this points to the gammainc_der op being an issue more than the parameterization

Reply:
> update: just looked at the [log-cdf of the gamma distribution](https://github.com/pymc-devs/pymc/blob/e64942be0bc643b2ffb709213546f3d0243bb5f7/pymc/distributions/continuous.py#L2320) and its implemented very similar to how I implemented the generalized gamma, so makes sense its slow - I guess this points to the gammainc_der op being an issue more than the parameterization
> 

Yes that's quite possible

Reply:
Given that fitting with the distribution works well without the Censored API and the regular Gamma distribution has the same problem in PyMC anyway, is it fair to call this ok and proceed with the PR (once tests are passing) since its experimental anyway?

It'd atleast be a good starting point to be improved upon, and the current parameterization I chose is fairly interpretable, and I still have some use cases for it atleast

Reply:
Definitely! I'll try to review it again soon.